verification/spec/relayer/ClientHandlers.tla

```diff
@@ -0,0 +1,94 @@
+
+CONSTANTS MaxHeight \* maximal height of all the chains in the system
+
+ClientIDs == {"clA", "clB"}
+nullClientID == "none"
+nullHeight == 0
+
+Heights == 1..MaxHeight 
+
+Max(S) == CHOOSE x \in S: \A y \in S: y <= x
+
+(***************************************************************************
+ Client helper operators
+ ***************************************************************************)
+
+\* get the ID of chainID's counterparty chain    
+GetCounterpartyChainID(chainID) ==
+    IF chainID = "chainA" THEN "chainB" ELSE "chainA"    
+ 
+\* get the client ID of the client for chainID 
+GetClientID(chainID) ==
+    IF chainID = "chainA" THEN "clA" ELSE "clB"
+        
+\* get the client ID of the client for chainID's counterparty chain           
+GetCounterpartyClientID(chainID) ==
+    IF chainID = "chainA" THEN "clB" ELSE "clA" 
+    
+
+(***************************************************************************
+ Client datagram handlers
+ ***************************************************************************)
+    
+\* Handle "CreateClient" datagrams
+HandleCreateClient(chainID, chain, datagrams) == 
+    \* get "CreateClient" datagrams with valid clientID
+    LET createClientDgrs == {dgr \in datagrams : 
+                            /\ dgr.type = "CreateClient"
+                            /\ dgr.clientID = GetCounterpartyClientID(chainID)} IN
+    \* get heights in datagrams with correct counterparty clientID for chainID
+    LET createClientHeights == {h \in Heights : \E dgr \in createClientDgrs : dgr.height = h} IN  
+    
+    \* new chain record with clients created
+    LET clientCreateChain == [
+            height |-> chain.height,
+            counterpartyClientHeight |-> IF createClientHeights /= {}
+                                         THEN Max(createClientHeights)
+                                         ELSE chain.counterpartyClientHeight,
+            connectionEnd |-> chain.connectionEnd
+         ] IN
+
+    IF chain.counterpartyClientHeight = nullHeight
+    \* if the counterparty client height is null, create the client   
+    THEN clientCreateChain
+    \* otherwise, do not update the chain  
+    ELSE chain 
+
+\* Handle "ClientUpdate" datagrams
+HandleUpdateClient(chainID, chain, datagrams) ==      
+    \* get "ClientUpdate" datagrams with valid clientID
+    LET updateClientDgrs == {dgr \in datagrams : 
+                            /\ dgr.type = "ClientUpdate"
+                            /\ dgr.clientID = GetCounterpartyClientID(chainID)} IN
+    \* get heights in datagrams with correct counterparty clientID for chainID
+    LET updateClientHeights == {h \in Heights : \E dgr \in updateClientDgrs : dgr.height = h} IN    
+
+    \* new chain record with clients updated
+    LET clientUpdatedChain == [
+            height |-> chain.height,
+            counterpartyClientHeight |-> IF updateClientHeights /= {}
+                                         THEN Max(updateClientHeights)
```

Comment:
We should store all installed heights, not just the maximum height.

Reply:
Indeed, installing only the maximum height violates liveness.